### PR TITLE
Allow tests to customize FixAllContext

### PR DESCRIPTION
**Pull request by tmat**
... and pass in internal options.

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs [hunk 1]
```diff
@@ -236,6 +236,54 @@ protected CodeFixTest()
...
+            FixAllScope scope,
+            string? codeActionEquivalenceKey,
+            IEnumerable<string> diagnosticIds,
+            FixAllContext.DiagnosticProvider fixAllDiagnosticProvider,
+            CancellationToken cancellationToken)
+            => new FixAllContext(document, codeFixProvider, scope, codeActionEquivalenceKey, diagnosticIds, fixAllDiagnosticProvider, cancellationToken);
+
+        /// <summary>
+        /// Creates a new <see cref="FixAllContext"/>.
+        /// Use this overload when applying fix all to a diagnostic with no source location, i.e. <see cref="Location.None"/>.
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs:265]: ❔ Is this even possible?

**Review by sharwell — CHANGES_REQUESTED** [review]
Only one `CreateFixAllContext` method should be virtual. If other overloads exist, they should ultimately delegate to the single virtual method.

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs (hunk 1 above)
Comment (tmat) [reply src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs:265]: The comment is copied from public API: https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/FixAllOccurrences/FixAllContext.cs,71

**Comment by tmat** [comment]
Remove the second overload.

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs [hunk 2]
```diff
@@ -778,7 +805,7 @@ private async Task<(Project project, ExceptionDispatchInfo? iterationCountFailur
                 var compilerDiagnosticIds = codeFixProviders.SelectMany(codeFixProvider => codeFixProvider.FixableDiagnosticIds).Where(x => x.StartsWith("CS", StringComparison.Ordinal) || x.StartsWith("BC", StringComparison.Ordinal));
                 var disabledDiagnosticIds = project.CompilationOptions.SpecificDiagnosticOptions.Where(x => x.Value == ReportDiagnostic.Suppress).Select(x => x.Key);
                 var relevantIds = analyzerDiagnosticIds.Concat(compilerDiagnosticIds).Except(disabledDiagnosticIds).Distinct();
-                var fixAllContext = new FixAllContext(fixableDocument, effectiveCodeFixProvider, scope, equivalenceKey, relevantIds, fixAllDiagnosticProvider, cancellationToken);
+                var fixAllContext = CreateFixAllContext(fixableDocument, effectiveCodeFixProvider!, scope, equivalenceKey, relevantIds, fixAllDiagnosticProvider, cancellationToken);
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs:808]: Should this be passing the project as well?

**Review by sharwell — APPROVED** [review]

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs (hunk 2 above)
Comment (tmat) [reply src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs]: Of course.

**Pull request merged**